### ISSUE TITLE
IRC utf8 message truncation yields might yield unwanted default decoding behavior

### ISSUE THREAD

**Issue by petercommand** [weechat/weechat#508]
As there is a message length restriction in irc, submitting a message that is over the message length restriction results in message truncation.
When the message contains UTF-8 characters, this might result in a string that is only a partially valid UTF-8 string, with the last UTF-8 character truncated.
As such, this string will not pass the function utf8_is_valid() called from string_iconv_to_internal(), resulting in falling back to using charset.default.decode to decode, which is an unwanted behavior in this case.
Setting charset.default.decode to empty string resolves this problem, however, this default behavior should be fixed.

**Comment by flashcode** [comment]
WeeChat has a "smart" cut of long messages, so it should never cut inside an UTF-8 char, but always before an UTF-8 char (so messages cut can be a little bit smaller than expected).
If this happens, please tell me a way to reproduce, with a test message for example (with the same headers in message, ie the host, nick, target channel, ...).

**Comment by petercommand** [comment]
It's not that weechat is cutting inside utf-8 char. but that the message received by weechat might have a truncated UTF-8 char.

**Comment by flashcode** [comment]
Ah I see. So I already answered that some years ago, my reply was that I don't know what I can do for that in WeeChat. The other clients truncating inside an UTF-8 char should be fixed.
I'm not sure how WeeChat could be sure there's an invalid UTF-8 char just at the end, this would be weird to make a special case for that.
If really needed, you could write a script that strips this char when the last one is invalid (or replaces it with some other char).

**Comment by petercommand** [comment]
WeeChat don't need to know that there's an invalid UTF-8 char just at the end. It just need to have a consistent default decoding policy for user, so that it uses the same decoding charset even for the invalid utf-8 string by default.
Sudden change in decoding is abnormal in normal chat rooms, so I think WeeChat should cancel the fallback decoding by default.

**Comment by flashcode** [comment]
I don't agree with that. Many users may send messages with different encodings in the same channel, I see no problem with that (I'm on such channels). So using a default charset "iso-8859-1" with default as "utf-8" is a feature, not a bug.
For me, the problem is the malformed messages, not the double decoding by default in WeeChat.

To fix your problem, you can just disable the default iso decoding (and enable it only on some channels, if needed): `/set charset.default.decode ""`

**Comment by flashcode** [comment]
@petercommand : removing the default decoding charset helped to fix your problem?
If yes I'll just mark this issue as "won't fix", since for me the bug is in other clients.

**Comment by petercommand** [comment]
Yes, removing the default decoding charset helps.
Seems that if the client doesn't truncate the message, the server will do the truncation though.

**Comment by flashcode** [comment]
So one solution would be to add a new option in WeeChat, to allow messages that are full UTF-8 compliant, except the last char (in case another client/server made a split inside an UTF-8 char).
So that it would not fallback on ISO-8859-1 in that case, and display "?" as last char.
